src/Sentry.Unity/Il2CppEventProcessor.cs

```diff
@@ -56,47 +54,163 @@ public void Process(Exception incomingException, SentryEvent sentryEvent)
+                        frame.InstructionAddress = $"0x{instructionAddress:X8}";
+                    }
+
+                    _ = usedImages.Add(image);
+                }
+            }
+
+            sentryEvent.DebugImages ??= new List<DebugImage>();
+            sentryEvent.DebugImages.AddRange(usedImages);
+        }
+
+        // Normalizes Debug Image UUID so that we can compare the ones coming from
+        // native (contains dashes, all lower-case) & what Unity gives us (no dashes, uppercase).
+        // On Linux, the image also has shorter UUID coming from Unity, e.g. 3028cb80b0712541,
+        // while native image UUID we get is 3028cb80-b071-2541-0000-000000000000.
+        private static string? NormalizeUUID(string? value) =>
+            value?.ToLowerInvariant().Replace("-", "").TrimEnd(new char[] { '0' });
+
+        private class DebugImageInfo
+        {
+            public readonly DebugImage Image;
+            public readonly ulong StartAddress;
+            public readonly ulong EndAddress;
+
+            public DebugImageInfo(DebugImage image)
+            {
+                Image = image;
+                StartAddress = Convert.ToUInt64(image.ImageAddress, 16);
+                EndAddress = StartAddress + (ulong)image.ImageSize!;
+            }
+        }
+
+        private static IDiagnosticLogger? _logger;
+
+        private static Lazy<List<DebugImageInfo>> DebugImagesSorted = new(() =>
+        {
+            var result = new List<DebugImageInfo>();
+            var nativeDebugImages = C.DebugImages.Value;
+            foreach (var image in nativeDebugImages)
+            {
+                if (image.ImageSize is null)
+                {
+                    _logger?.Log(SentryLevel.Debug,
+                        "Skipping debug image '{0}' (CodeId {1} | DebugId: {2}) because its size is NULL",
+                        null, image.CodeFile, image.CodeId, image.DebugId);
+                    continue;
+                }
+
+                var info = new DebugImageInfo(image);
+                int i = 0;
+                for (; i < result.Count; i++)
+                {
+                    if (info.StartAddress < result[i].StartAddress)
+                    {
+                        // insert at index `i`, all the rest have a larger start address
+                        break;
+                    }
+                }
+                result.Insert(i, info);
+
+                _logger?.Log(SentryLevel.Debug,
+                    "Found debug image '{0}' (CodeId {1} | DebugId: {2}) with addresses between {3:X8} and {4:X8}",
+                    null, image.CodeFile, image.CodeId, image.DebugId, info.StartAddress, info.EndAddress);
+            }
+            return result;
+        });
+
+        private static DebugImage? FindDebugImageContainingAddress(ulong instructionAddress)
+        {
+            var list = DebugImagesSorted.Value;
+            foreach (var info in list)
```

Comment:
doesn’t C# have a binary search in the stdlib? iterating the whole list for each frame may be expensive

Reply:
Changed to binary search, though I think it's a bit less readable